diff -r -N -U0 larpix_network_sim/larpix_v3c/src/comms_ctrl.sv larpix_network_sim/larpix_v3c2/src/comms_ctrl.sv
--- larpix_network_sim/larpix_v3c/src/comms_ctrl.sv	2026-04-29 20:26:56.175315000 -0700
+++ larpix_network_sim/larpix_v3c2/src/comms_ctrl.sv	2026-05-06 10:50:51.668968000 -0700
@@ -141,2 +141,2 @@
-always_comb 
-    pkt_data = (pkt_chip == chip_id) ? {~^read_pkt,read_pkt} : rx_data;
+always_comb
+    pkt_data = {~^read_pkt, read_pkt};
@@ -238,4 +238,4 @@
-                if ((rx_data[9:2] != chip_id) || 
-                        (rx_data[9:2] == GLOBAL_ID) ||
-                        pkt_type == DATA_OP) begin
-                        read_pkt <= rx_data;
+                if ((rcvd_pkt[9:2] != chip_id) || 
+                        (rcvd_pkt[9:2] == GLOBAL_ID) ||
+                        rcvd_pkt[1:0] == DATA_OP) begin
+                        read_pkt <= rcvd_pkt[WIDTH-2:0];
diff -r -N -U0 larpix_network_sim/larpix_v3c/src/hydra_ctrl.sv larpix_network_sim/larpix_v3c2/src/hydra_ctrl.sv
--- larpix_network_sim/larpix_v3c/src/hydra_ctrl.sv	2026-04-30 13:54:07.626848000 -0700
+++ larpix_network_sim/larpix_v3c2/src/hydra_ctrl.sv	2026-05-06 10:53:46.089935000 -0700
@@ -159 +159,2 @@
-            else if (!fifo_empty)            Next = TX_GET_FIFO;
+            else if (!fifo_empty && ((tx_busy & enable_piso_downstream) == '0))
+                                            Next = TX_GET_FIFO;
@@ -172,2 +173 @@
-                else if (|uart_has_data)    Next = RX_CAPTURE;
-            else                            Next = IDLE;
+                else                        Next = IDLE;
@@ -176 +176 @@
-            if ((tx_busy & enable_piso_downstream) != '0) Next = TX_GET_FIFO; 
+            if ((tx_busy & enable_piso_downstream) != '0) Next = IDLE; 
