diff -r -N -U0 larpix_network_sim/larpix_v3b_rtl_v2/src/comms_ctrl.sv larpix_network_sim/larpix_v3b_rtl/src/comms_ctrl.sv
--- larpix_network_sim/larpix_v3b_rtl_v2/src/comms_ctrl.sv	2026-04-29 19:57:39.047306000 -0700
+++ larpix_network_sim/larpix_v3b_rtl/src/comms_ctrl.sv	2026-04-29 19:57:38.491303000 -0700
@@ -130,2 +130,2 @@
-always_comb
-    pkt_data = {~^read_pkt, read_pkt};
+always_comb 
+    pkt_data = (pkt_chip == chip_id) ? {~^read_pkt,read_pkt} : rx_data;
@@ -227,4 +227,4 @@
-                if ((rcvd_pkt[9:2] != chip_id) || 
-                        (rcvd_pkt[9:2] == GLOBAL_ID) ||
-                        rcvd_pkt[1:0] == DATA_OP) begin
-                        read_pkt <= rcvd_pkt[WIDTH-2:0];
+                if ((rx_data[9:2] != chip_id) || 
+                        (rx_data[9:2] == GLOBAL_ID) ||
+                        pkt_type == DATA_OP) begin
+                        read_pkt <= rx_data;
diff -r -N -U0 larpix_network_sim/larpix_v3b_rtl_v2/src/hydra_ctrl.sv larpix_network_sim/larpix_v3b_rtl/src/hydra_ctrl.sv
--- larpix_network_sim/larpix_v3b_rtl_v2/src/hydra_ctrl.sv	2026-04-29 19:57:39.082302000 -0700
+++ larpix_network_sim/larpix_v3b_rtl/src/hydra_ctrl.sv	2026-04-29 19:57:38.526303000 -0700
@@ -135,0 +136,2 @@
+
+
@@ -154,2 +156 @@
-            else if (!fifo_empty && ((tx_busy & enable_piso_downstream) == '0))
-                                            Next = TX_GET_FIFO;
+            else if (!fifo_empty)            Next = TX_GET_FIFO;
@@ -168 +169,2 @@
-                else                        Next = IDLE;
+                else if (|uart_has_data)    Next = RX_CAPTURE;
+            else                            Next = IDLE;
@@ -171 +173 @@
-            if ((tx_busy & enable_piso_downstream) != '0) Next = IDLE; 
+            if ((tx_busy & enable_piso_downstream) != '0) Next = TX_GET_FIFO; 
@@ -195,8 +197,10 @@
-            IDLE:
-                case (sel_onehot)
-                    4'b0001 : rx_data <= rx_data_uart[2'b00];
-                    4'b0010 : rx_data <= rx_data_uart[2'b01];
-                    4'b0100 : rx_data <= rx_data_uart[2'b10];
-                    4'b1000 : rx_data <= rx_data_uart[2'b11];
-                    default : rx_data <= rx_data;
-                endcase
+            IDLE: 
+                if (|uart_has_data)
+                    case (sel_onehot)
+                        4'b0001 : rx_data <= rx_data_uart[2'b00]; 
+                        4'b0010 : rx_data <= rx_data_uart[2'b01]; 
+                        4'b0100 : rx_data <= rx_data_uart[2'b10]; 
+                        4'b1000 : rx_data <= rx_data_uart[2'b11]; 
+                        default : rx_data <= '0;
+                    endcase
+
